File: waterball/ch1/practices/p1_dependency/py/hero/hero.py
from .levelsheet import LevelSheet


class Hero:
    def __init__(self):
        self._level = 1
        self._total_exp = 0
        self._hp = 100

    def gain_exp(self, exp: int, level_sheet: LevelSheet):
        if exp < 0:
            raise ValueError(f"error: 無法獲得負數的經驗值 {exp}")
        current_level = self.level
        self.total_exp = self.total_exp + exp
        self.level = level_sheet.query_level(self.total_exp)
        print(
            f"英雄目前等級 {current_level}，獲得 {exp} EXP，最新總共經驗值為 {self.total_exp}，最新等級為 {self.level}"
        )

    @property
    def hp(self) -> int:
        return self._hp

    @hp.setter
    def hp(self, new_hp: int):
        if new_hp < 0:
            raise ValueError(f"HP must be greater than or equal to 0, got {new_hp}")
        self._hp = new_hp

    @property
    def total_exp(self) -> int:
        return self._total_exp

    @total_exp.setter
    def total_exp(self, exp: int):
        if exp < 0:
            raise ValueError(f"Total EXP must be greater than or equal to 0, got {exp}")
        self._total_exp = exp

    @property
    def level(self) -> int:
        return self._level

    @level.setter
    def level(self, new_level: int):
        if new_level < 0:
            raise ValueError(
                f"Level must be greater than or equal to 0, got {new_level}"
            )
        self._level = new_level

File: waterball/ch1/practices/p1_dependency/golang/hero/hero.go
package hero

import (
	"fmt"
)

type Hero struct {
	level    int
	totalExp int
	hp       int
}

// NewHero creates a new Hero with default values.
func NewHero() *Hero {
	return &Hero{
		level:    1,
		totalExp: 0,
		hp:       100,
	}
}

func (h *Hero) GainExp(exp int, levelSheet LevelSheet) error {
	if exp < 0 {
		return fmt.Errorf("error: 無法獲得負數的經驗值 %d", exp)
	}
	currentLevel := h.getLevel()
	h.setTotalExp(h.getTotalExp() + exp)
	h.setLevel(levelSheet.QueryLevel((h.getTotalExp())))
	fmt.Printf("英雄目前等級 %d，獲得 %d EXP，最新總共經驗值為 %d，最新等級為 %d。\n", currentLevel, exp, h.getTotalExp(), h.getLevel())
	return nil
}

// setTotalExp sets the total experience points, ensuring it's non-negative.
func (h *Hero) setTotalExp(totalExp int) error {
	if totalExp < 0 {
		return fmt.Errorf("TotalExp must be greater than or equal to 0, got %d", totalExp)
	}
	h.totalExp = totalExp
	return nil
}

// setLevel sets the level, ensuring it's non-negative.
func (h *Hero) setLevel(level int) error {
	if level < 0 {
		return fmt.Errorf("Level must be greater than or equal to 0, got %d", level)
	}
	h.level = level
	return nil
}

// setLevel sets the hp, ensuring it's non-negative.
func (h *Hero) setHp(hp int) error {
	if hp < 0 {
		return fmt.Errorf("HP must be greater than or equal to 0, got %d", hp)
	}
	h.hp = hp
	return nil
}

func (h Hero) getTotalExp() int {
	return h.totalExp
}

func (h Hero) getLevel() int {
	return h.level
}

func (h Hero) getHp() int {
	return h.hp
}
